fmt print statements replaced using the log package

10eMyrT opened the pull request:
Closes #245.

Good evening! I went through what we already have, and replaced the print statements generated with`fmt` by `log`. However, I noticed two issues, I would like your thoughts on.
* In the _**handlers.go**_ function, replacing the use of `fmt` in this type of statement `c.sendToTg(fmt.Sprintf(partFmt, e.Source.Name))` has not worked with `log` because the later writes to a stream, whereas the statement writes to a char array with fmt. An alternative to `fmt` would be to use `strings.Builder` but I find it more complex than the current statement.

* Adopting levels like _All, Useful_, would require the change of the flag setup from bool to int, and might require passing the debug level argument when calling `go ircClient.StartBot(ircChan, tgClient.SendMessage)` and `go tgClient.StartBot(tgChan, ircClient.SendMessage)`. To my current knowledge of golang, it would mean the use of if-else statements to determine whether to print or not.

Tjzabel commented:
@10eMyrT in regards to your first comment, the `fmt` here is formatting the string to send over to Telegram. This should not be touched by any logging.

If possible, I would like us to be able to instantiate a `log` object only once or twice, and not in every file the logging is needed. Commonly, this would mean adding a `log` object within our `Client` structs. However, this is a good first start. 

I will add my definitive review once I can get around to testing how this logging works :+1:

Zedjones commented on internal/handlers/irc/irc.go:
```diff
@@ -17,11 +18,20 @@ type Client struct {
+var (
+	logFlags    = log.Ldate | log.Ltime | log.Lmicroseconds | log.Llongfile
+	Info        = log.New(os.Stdout, "INFO: ", logFlags)
+	Error       = log.New(os.Stderr, "ERROR: ", logFlags)
```
At the moment, this causes the code not to compile. You need to import `os` as well to use it's namespace. This also applies to telegram.go.

Zedjones replied:
Oh, also, since Go has syntactic capitalization, these variables are now both exported from this module, which is not desired. Please change these to `info` and `error`, respectively.

Zedjones commented on internal/handlers/irc/irc.go:
```diff
@@ -17,11 +18,20 @@ type Client struct {
+Creating variables for logging
+*/
+var (
+	logFlags    = log.Ldate | log.Ltime | log.Lmicroseconds | log.Llongfile
```
Could you please switch instances of `Llongfile` to `Lshortfile`? I think having the fully qualified name is most unnecessary here, since most deployments will use something like systemd or containers and the full path wouldn't be very helpful in these situatoins.

Zedjones commented on internal/handlers/irc/irc.go:
```diff
@@ -1,7 +1,8 @@
 import (
-	"fmt"
+	"log"
+	"io/ioutil"
```
Unused import here, as well as the other files that import this.

10eMyrT replied:
Thanks for the comments. I have considered Tim's and your feedback to reorganise. I should have something for today's meeting.

Zedjones commented:
As Tim was saying, it would be a good idea to only have two logger objects (one for error and one for info) and store them in the respective client structs. Or, even better, have a logger struct that contains methods for "logInfo" and "logError" which handles calling the correct logger's `Println` method. 

Also, just running a `go build cmd/teleirc.go` is a good way to make sure that there are no compiler errors/warnings before committing in the future (since we currently don't have any CI set up for Go).

10eMyrT commented:
I considered the feedback and modified the code. Note that in its current format, the print statements are called in `debug.go`. Therefore, I removed the `Lshortfile` logFlag because it would always print `debug.go`. I also guess knowing the file is not important for INFO. It might be for WARNING and ERROR. There is the [runtime package](https://golang.org/pkg/runtime/) which allows to get the caller function, the filePath would be a long string starting from `/home/......`.

Zedjones commented on internal/debug.go:
```diff
@@ -0,0 +1,42 @@
+	errorLog    = log.New(os.Stderr, "ERROR: ", logFlags)
+	warning     = log.New(os.Stderr, "WARNING: ", logFlags)
+)
+
```
Could you add an interface that defines these functions, something like:
```go
type DebugLogger interface {
	LogInfo(string)
	LogError(error)
	LogWarning(string)
	PrintVersion(string)
}
```
This will help a lot with testing later on.

Zedjones commented on internal/handlers/irc/irc.go:
```diff
@@ -14,14 +13,15 @@ and the IRCSettings that were passed into NewClient
 type Client struct {
 	*girc.Client
 	Settings internal.IRCSettings
+	verbose internal.Debug
```
Once you define that interface, please change structs to contain an `internal.DebugLogger` (or whatever you call it) instead of `internal.Debug`

Zedjones commented on internal/handlers/irc/irc.go:
```diff
@@ -14,14 +13,15 @@ and the IRCSettings that were passed into NewClient
 */
-func NewClient(settings internal.IRCSettings) Client {
-	fmt.Println("Creating new IRC bot client...")
+func NewClient(settings internal.IRCSettings, debug internal.Debug) Client {
```
You'll also need to change the `NewClient` function signatures to accept the interface instead

jwflory commented:
@Tjzabel @Zedjones Do you think this PR is ready to go?

Zedjones commented:
> @Tjzabel @Zedjones Do you think this PR is ready to go?

I've just got one last change to request. and it's a linting error. Assuming this has been rebased since the other PR got landed, it should be almost good to go.

Zedjones commented on cmd/teleirc.go:
```diff
@@ -18,42 +17,43 @@ const (
 func main() {
 	flag.Parse()
 
+	verbose := internal.Debug { *flagDebug	}
```
Could you change this to reference the field name? Not a huge deal, but it's a linting error for me.
```go
verbose := internal.Debug{DebugLevel: *flagDebug}
```

10eMyrT replied:
Sure. I already did it.

jwflory commented:
Awesome. Everything looks good to me. Going to merge this in! :ocean: Thanks for working on this one @10eMyrT. :tada:

Tjzabel commented:
I just saw these updates. I added my own review of this PR accidentally into #264
